### DISCUSSION TITLE
Development of some component system besides ActiveX

### DISCUSSION THREAD

**Discussion by YogiYang** [twinbasic/twinbasic#407] [General]
@WaynePhillipsEA ,

Are there any plans to implement some sort of component system besides ActiveX in tB at IDE level as well as compiler level?

I personally feel that this is required as it opens up a whole new industry of component developers.

This is the reason why VB6 competitors like Delphi, Visual Studio (.net), FPC with Lazarus, etc. are so successful and PureBasic, FreeBasic, etc. are not that successful in the market.

I know you may think ActiveX is already there, why build another component/control system. But when it comes to cross platform ActiveX will not work (on non windows platform) so a full fledged component system should be built that is guaranteed to work on all platforms.

I would advise you to please study Delphi and FPC with Lazarus as the component system they have implemented actually compiles the components right into the final EXE so one does not have to distribute them separately and this helps avoid version conflict hell as is the case currently in VB6 as well as in case of VS.net.

**Comment by WaynePhillipsEA** [comment]
Yes, we've got some vague plans in this area.   The initial support will likely be along the lines as detailed in https://github.com/WaynePhillipsEA/twinbasic/issues/16 (specifically https://github.com/WaynePhillipsEA/twinbasic/issues/16#issuecomment-867078538).

**Reply by YogiYang** [reply]
> 
> 
> Yes, we've got some vague plans in this area. The initial support will likely be along the lines as detailed in #16 (specifically [#16 (comment)](https://github.com/WaynePhillipsEA/twinbasic/issues/16#issuecomment-867078538)).

After reading the post I feel that supporting only source based component system will not attract many component developer, esp. commercial developers like ComponenOne, Desaware, etc. who sell their components.

Just another though can you add support for hosting .NET components. If this is possible then I feel it would be easy to attract and divert VB.net developers towards tB.

For example tools like WinDev support ActiveX as well as .Net (non-visual) components along with their own custom component format.

**Comment by Kr00l** [comment]
To note is that PureBasic and FreeBasic are not "COM based". So, it will be like day/night between those and tB.
Maybe that's also a reason why they are not so successful as they can't replace VBA/VB6.

**Reply by YogiYang** [reply]
I think replacing VB6 is not important any more outside the basic VBForum community. In the commercial world it is the supporting ecosystem that counts.

@Kr00l, As a developer do you really expect yourself or other developers to develop all required UI controls from scratch for their project. Of course not. And I think those competitors mentions in my first post have missed that part.

**Comment by YogiYang** [comment]
I think this post may give @WaynePhillipsEA some ideas on component designing and management.

https://www.patreon.com/posts/server-fixes-new-56026280

@WaynePhillipsEA, Just like you this guy is a one man army and it doing just great building web based OS as well as development IDE.!!!
